### ISSUE TITLE
Using hot reload for plain es6 components (non-functional)

### ISSUE THREAD

**Issue by jonny-novikov** [vuejs/vue-loader#461]
I just rewrote some of components (and changed .vue to .js)

    import "./sass/app.scss";
    const tpl = `<div class="app"></app>`; // with literals
    export default {
      template: tpl,
      data: { ... }
      ...
    }

But writing in this manner I don't touch vue-loader and got full page reload.
This approach gives well-organized code splitting (js/html) in Webstorm.

@yyx990803 Is it possible to enable hot reload for vanilla-js Vue opts?
And is it a good approach to declare components?

**Comment by michaelkrupp** [comment]
I'd be interested in this as well as I really do not like putting template, style and logic together into a single file - aside from very poor editor/linter support atm.

I already found [vue-hot-reload-api] and the implementation in [template-compiler] and [loader], but it looks like it is non-trivial to do without a heap of boilerplate-code and/or a custom template loader.

[vue-hot-reload-api]: https://github.com/vuejs/vue-hot-reload-api
[template-compiler]: https://github.com/vuejs/vue-loader/blob/master/lib/template-compiler.js#L80
[loader]: https://github.com/vuejs/vue-loader/blob/master/lib/loader.js#L286

**Comment by jonny-novikov** [comment]
@michaelkrupp Yes, it's really non-trivial for me too.
I started to investigate with code like:

    if (module.hot) {
      // Whenever a new version of App.js is available
      module.hot.accept('./App', function () {
      const api = require('vue-hot-reload-api');
      ..... 

but I gave it up because I use state management and it's not a fail to loose state.
But putting template in a separate file is not a good idea.

The main goal is to keep "vanilla" file and even there are no Vue dependencies and so on.
Opts with template can be loaded everywhere, it's obvious.

But of cource I can do linting for vue files too but it has a lot of lacks in most cases.

**Comment by yyx990803** [comment]
1. You can split up a `.vue` file and get hot reload like usual:

``` html
<style src="./sass/app.scss" lang="scss"></style>
<script src="./app.js"></script>
<template>
  ...
</template>
```

2. It's possible to write a separate `vue-hot-loader` using `vue-hot-reload-api`, but I don't think I have time to do that anytime soon. Feel free to give it a shot though.

**Comment by jonny-novikov** [comment]
@yyx990803 Evan, thank you for the anwer. I closed the related issue in vue-hot-reload-api.

Yeap. I invest some time and tried to solve it with approach like:

project structure:

    src/
      App.js
      App.vue

and App.vue contains something like this:

    <template>...</template>
    <script src="./app.js"></script>

And I have got an ability to hot reload the app (because HMR know the dependency tree).
All other vue components I still can place as js files in src/components/**.js

But the goal was **not to split** into separate file, but to write "vanilla-js" dep-free opts es6 file with template string. That's why I can still keep template and javascript in a single file (splitting it automaticaly when 'Edit HTML template').

Results:

* I have full refactoring tools
* Full linting support
* Clean mind ;)

BTW, it's awesome to write Vue components! Thank you!
